nixos/modules/security/pass.nix

```diff
@@ -0,0 +1,200 @@
+          default = "root";
+          description = ''
+            Owner of the secret <literal>path</literal>.
+          '';
+        };
+        group = lib.mkOption {
+          type = types.str;
+          default = "root";
+          description = ''
+            Group of the secret <literal>path</literal>.
+          '';
+        };
+        pipe = lib.mkOption {
+          type = types.nullOr types.str;
+          default = null;
+          description = ''
+            Shell command taking the deciphered secret on its standard input
+            and which must put on its standard output
+            the actual material to be installed.
+            This allows to decorate the secret with non-secret bits.
+          '';
+        };
+        path = lib.mkOption {
+          type = types.str;
+          default = name;
+          apply = p: if match "^/.*" p == null then "/run/pass-secrets/"+p+"/file" else p;
+          description = ''
+            The path on the target system where the secret is installed to.
+            Any non-absolute path is relative to <filename>/run/pass-secrets</filename>.
+            Default to the name of the secret.
+          '';
+        };
+        service = lib.mkOption {
+          type = types.str;
+          default = "secret-" + escapeUnitName name + ".service";
+          description = ''
+            The name of the systemd service.
+            Useful to put constraints like <literal>after</literal> or <literal>wants</wants>
+            into services requiring this secret.
+          '';
+        };
+        postStart = lib.mkOption {
+          type = types.lines;
+          default = "";
+          example = "systemctl reload nginx.service";
+          description = ''
+            Commands to run after new secrets go live.
+            Typically the web server and other servers using secrets
+            need to be reloaded.
+          '';
+        };
+        postStop = lib.mkOption {
+          type = types.lines;
+          default = "";
+          example = ''shred -u "$secret_file"'';
+          description = ''
+            Commands to run after stopping the service.
+            Typically removing a persistent secret.
+          '';
+        };
+      };
+    }));
+  };
+};
+config = lib.mkIf (pass.secrets != {}) {
+  systemd.services =
+    lib.mapAttrs' (target: secret:
+      lib.nameValuePair (lib.removeSuffix ".service" secret.service) {
+        description = "Install secret ${secret.path}";
+        script = ''
+          set -o pipefail
+          set -eux
+          decrypt() {
+            {
+            ${pkgs.gnupg}/bin/gpg --batch --pinentry-mode loopback \
+             --passphrase-file '${secret.passphraseFile}' \
+             --decrypt '${secret.gpg}' \
+            ${lib.optionalString (secret.pipe != null) (" | "+secret.pipe)}
+            } |
+            install -D -m '${secret.mode}' -o '${secret.user}' -g '${secret.group}' /dev/stdin \
```

Comment:
What do you think about installing it to `/run/keys` by default. In sops-nix I do this and create symlinks from their. This way secrets are kept in RAM. Combined with hardware tokens for GPG no secret would be persistently stored.

Reply:
By default secrets are already installed in `/run`, and now the passphrase no longer comes from a file but only from `gpg-agent`. This passphrase can be uploaded each time with `gpg-preset-passphrase`, or made persistent in a file loaded with `gpg-preset-passphrase` in the `postStart` of `gpg-agent.service`.

Reply:
I mean `/run/keys` not `/run/` because tmpfs can be swapped to disk. But we also discussed this here: https://github.com/NixOS/nixpkgs/pull/93659#discussion_r459849419